[JSC] Add CachedCall fast path for specific argument count

Constellation opened the pull request:
#### 240fb30df0de1f3b18ee5a4ca0c8c3b683044d50
<pre>
[JSC] Add CachedCall fast path for specific argument count
<a href="https://bugs.webkit.org/show_bug.cgi?id=272860">https://bugs.webkit.org/show_bug.cgi?id=272860</a>
<a href="https://rdar.apple.com/126651919">rdar://126651919</a>

Reviewed by Keith Miller.

Let&apos;s make up-to-3 argumentCount CachedCall incredibly faster. We would like to hand-craft that for the other argument counts.
Based on static argument count, we use super fast asm setup in LLInt to invoke JS function.

                                        ToT                     Patched

    uint32array-sort-custom       73.6850+-0.7307     ^     53.4209+-1.7978        ^ definitely 1.3793x faster

* JSTests/microbenchmarks/uint32array-sort-custom.js: Added.
(i.array.sort):
* Source/JavaScriptCore/interpreter/CachedCall.h:
(JSC::CachedCall::CachedCall):
(JSC::CachedCall::callWithArguments):
* Source/JavaScriptCore/interpreter/Interpreter.h:
* Source/JavaScriptCore/interpreter/InterpreterInlines.h:
(JSC::Interpreter::tryCallWithArguments2):
* Source/JavaScriptCore/llint/LLIntThunks.cpp:
* Source/JavaScriptCore/llint/LLIntThunks.h:
* Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
* Source/JavaScriptCore/llint/LowLevelInterpreter64.asm:
* Source/JavaScriptCore/offlineasm/arm64.rb:
* Source/JavaScriptCore/offlineasm/registers.rb:
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeFunctions.h:
(JSC::genericTypedArrayViewProtoFuncSortImpl):
* Source/JavaScriptCore/runtime/StringPrototypeInlines.h:
(JSC::replaceUsingStringSearch):

Canonical link: <a href="https://commits.webkit.org/277699@main">https://commits.webkit.org/277699@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/028c6930f87d5a9950bdf180df42c1b38de1a9bf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48312 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27524 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/51266 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51000 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44377 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/33460 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25045 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/39476 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48894 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25220 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/41735 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20619 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22700 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/42915 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6368 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/41609 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/44658 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/43373 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52904 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/47803 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/23359 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/19709 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24624 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/41922 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25429 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/55298 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6872 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/24347 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11378 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/interpreter/InterpreterInlines.h:
```diff
@@ -107,20 +107,60 @@ ALWAYS_INLINE JSValue Interpreter::executeCachedCall(CachedCall& cachedCall)
+    else if constexpr (sizeof...(args) == 3)
+        return JSValue::decode(vmEntryToJavaScriptWith3Arguments(entry, &vm, codeBlock, callee, thisValue, args...));
+    else
+        return { };
```
This should be unreachable right? Can we change it to one of `RELEASE_ASSERT_NOT_REACHED()` or `static_assert(false)`?

Constellation replied:
I think `return { };` is fine since the caller will just fall back to the normal path when it returns JSEmpty value.

kmiller68 commented on Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
```diff
@@ -1694,6 +1694,76 @@ else
+        pushCalleeSaves()
+        vmEntryRecord(cfr, sp)
+        storep a1, VMEntryRecord::m_vm[sp]
+        loadpairq VM::topCallFrame[a1], t8, t9
```
I think it's good practice to put a comment that t9 is topEntryFrame for readability.

Constellation replied:
Done.

kmiller68 commented on Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
```diff
@@ -1694,6 +1694,76 @@ else
+        # entry must be a0
+        vmEntryToJavaScriptSetup()
+        move 1, t8 # argumentCountIncludingThis
+        subp 48, sp
```
Can we make this not a magic number?

Constellation replied:
Done.

kmiller68 commented on Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
```diff
@@ -1694,6 +1694,76 @@ else
+        # entry must be a0
+        vmEntryToJavaScriptSetup()
+        move 2, t8 # argumentCountIncludingThis
+        subp 64, sp
```
Ditto.

Constellation replied:
Done.

kmiller68 commented on Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
```diff
@@ -1694,6 +1694,76 @@ else
+        vmEntryToJavaScriptSetup()
+        move 2, t8 # argumentCountIncludingThis
+        subp 64, sp
+        storepairq a2, a3, CodeBlock + (8 * 0)[sp]
```
CodeBlock + (SlotSize * 0)

Constellation replied:
Done.

kmiller68 commented on Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
```diff
@@ -1694,6 +1694,76 @@ else
+        subp 64, sp
+        storepairq a2, a3, CodeBlock + (8 * 0)[sp]
+        storepairq t8, a4, CodeBlock + (8 * 2)[sp]
+        storepairq a5, a5, CodeBlock + (8 * 4)[sp]
```
Ditto.

Constellation replied:
Done.

kmiller68 commented on Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
```diff
@@ -1694,6 +1694,76 @@ else
+        storepairq t8, a4, CodeBlock + (8 * 2)[sp]
+        storepairq a5, a5, CodeBlock + (8 * 4)[sp]
+        move sp, t8
+        storepairq t8, cfr, VM::topCallFrame[a1]
```
Ditto on comment saying what we're storing.

Constellation replied:
Done.

kmiller68 commented on Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
```diff
@@ -1694,6 +1694,76 @@ else
+        # entry must be a0
+        vmEntryToJavaScriptSetup()
+        move 3, t8 # argumentCountIncludingThis
+        subp 64, sp
```
Ditto.

Constellation replied:
Done.

kmiller68 commented on Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
```diff
@@ -1694,6 +1694,76 @@ else
+        subp 64, sp
+        storepairq a2, a3, CodeBlock + (8 * 0)[sp]
+        storepairq t8, a4, CodeBlock + (8 * 2)[sp]
+        storepairq a5, a6, CodeBlock + (8 * 4)[sp]
```
Ditto.

Constellation replied:
Done.

kmiller68 commented on Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
```diff
@@ -1694,6 +1694,76 @@ else
+        storepairq t8, a4, CodeBlock + (8 * 2)[sp]
+        storepairq a5, a6, CodeBlock + (8 * 4)[sp]
+        move sp, t8
+        storepairq t8, cfr, VM::topCallFrame[a1]
```
Ditto.

Constellation replied:
Done.

kmiller68 commented on Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
```diff
@@ -1694,6 +1694,76 @@ else
+        # entry must be a0
+        vmEntryToJavaScriptSetup()
+        move 4, t8 # argumentCountIncludingThis
+        subp 80, sp
```
Ditto.

Constellation replied:
Done.

kmiller68 commented on Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
```diff
@@ -1694,6 +1694,76 @@ else
+        storepairq a2, a3, CodeBlock + (8 * 0)[sp]
+        storepairq t8, a4, CodeBlock + (8 * 2)[sp]
+        storepairq a5, a6, CodeBlock + (8 * 4)[sp]
+        storepairq a7, a7, CodeBlock + (8 * 6)[sp]
```
Ditto.

Constellation replied:
Done.

kmiller68 commented on Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
```diff
@@ -1694,6 +1694,76 @@ else
+        storepairq a5, a6, CodeBlock + (8 * 4)[sp]
+        storepairq a7, a7, CodeBlock + (8 * 6)[sp]
+        move sp, t8
+        storepairq t8, cfr, VM::topCallFrame[a1]
```
Ditto.

Constellation replied:
Done.

kmiller68 commented on Source/JavaScriptCore/llint/LowLevelInterpreter64.asm:
```diff
@@ -390,6 +390,7 @@ _llint_throw_stack_overflow_error_from_vm_entry:
 # a0, a2, t3, t4
 macro makeJavaScriptCall(entry, protoCallFrame, temp1, temp2)
+_llint_call_javascript:
     addp 16, sp
```
Nit: can we make this `addp sizeof(CallerFrameAndPC)`

Constellation replied:
Done.

kmiller68 commented on Source/JavaScriptCore/llint/LowLevelInterpreter64.asm:
```diff
@@ -429,6 +430,7 @@ macro makeHostFunctionCall(entry, protoCallFrame, temp1, temp2)
     end
 end
 
+
```
Revert please.

Constellation replied:
Done.

kmiller68 commented on Source/JavaScriptCore/interpreter/InterpreterInlines.h:
```diff
@@ -107,20 +107,60 @@ ALWAYS_INLINE JSValue Interpreter::executeCachedCall(CachedCall& cachedCall)
+        vm.didEnterVM = true;
+    });
+
+    // We don't handle `NonDebuggerAsyncEvents` explicitly here. This is JS function (since this is CachedCall),
```
Ditto.

Constellation replied:
Fixed.

kmiller68 commented on Source/JavaScriptCore/interpreter/InterpreterInlines.h:
```diff
@@ -107,20 +107,60 @@ ALWAYS_INLINE JSValue Interpreter::executeCachedCall(CachedCall& cachedCall)
-        if (vm.hasExceptionsAfterHandlingTraps())
-            return throwScope.exception();
-    }
+    // We don't handle `NonDebuggerAsyncEvents` explicitly here. This is JS function (since this is CachedCall),
```
Nit: This is JS function => This is a JSFunction

Constellation replied:
Fixed.

kmiller68 commented on Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
```diff
@@ -1694,6 +1694,76 @@ else
+    end
+
+    # EncodedJSValue vmEntryToJavaScriptWith0Arguments(void*, VM*, CodeBlock*, JSObject*, JSValue, JSValue)
+    # entry, vm, codeBlock, callee, thisValue, arg0, arg1
```
This comment is wrong. it's 0 args, so I think it should be:
```
    # EncodedJSValue vmEntryToJavaScriptWith0Arguments(void*, VM*, CodeBlock*, JSObject*, JSValue)
    # entry, vm, codeBlock, callee, thisValue
```

Constellation replied:
Fixed.

kmiller68 commented on Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
```diff
@@ -1694,6 +1694,76 @@ else
+        jmp _llint_call_javascript
+
+    # EncodedJSValue vmEntryToJavaScriptWith1Arguments(void*, VM*, CodeBlock*, JSObject*, JSValue, JSValue, JSValue)
+    # entry, vm, codeBlock, callee, thisValue, arg0, arg1
```
Ditto. 
```   
    # EncodedJSValue vmEntryToJavaScriptWith1Arguments(void*, VM*, CodeBlock*, JSObject*, JSValue, JSValue)
    # entry, vm, codeBlock, callee, thisValue, arg0
```

Constellation replied:
Fixed.

kmiller68 reviewed (approved):
r=me with comment fix.

Constellation committed:
[JSC] Add CachedCall fast path for specific argument count
https://bugs.webkit.org/show_bug.cgi?id=272860
rdar://126651919

Reviewed by Keith Miller.

Let's make up-to-3 argumentCount CachedCall incredibly faster. We would like to hand-craft that for the other argument counts.
Based on static argument count, we use super fast asm setup in LLInt to invoke JS function.

                                        ToT                     Patched

    uint32array-sort-custom       73.6850+-0.7307     ^     53.4209+-1.7978        ^ definitely 1.3793x faster

* JSTests/microbenchmarks/uint32array-sort-custom.js: Added.
(i.array.sort):
* Source/JavaScriptCore/interpreter/CachedCall.h:
(JSC::CachedCall::CachedCall):
(JSC::CachedCall::callWithArguments):
* Source/JavaScriptCore/interpreter/Interpreter.h:
* Source/JavaScriptCore/interpreter/InterpreterInlines.h:
(JSC::Interpreter::tryCallWithArguments2):
* Source/JavaScriptCore/llint/LLIntThunks.cpp:
* Source/JavaScriptCore/llint/LLIntThunks.h:
* Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
* Source/JavaScriptCore/llint/LowLevelInterpreter64.asm:
* Source/JavaScriptCore/offlineasm/arm64.rb:
* Source/JavaScriptCore/offlineasm/registers.rb:
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeFunctions.h:
(JSC::genericTypedArrayViewProtoFuncSortImpl):
* Source/JavaScriptCore/runtime/StringPrototypeInlines.h:
(JSC::replaceUsingStringSearch):

Canonical link: https://commits.webkit.org/277699@main
